src/dbnode/client/circuitbreaker/middleware/middleware.go

```diff
@@ -0,0 +1,245 @@
+package middleware
+
+import (
+	"github.com/m3db/m3/src/dbnode/client/circuitbreaker/circuitbreakererror"
+	"github.com/m3db/m3/src/dbnode/client/circuitbreaker/internal/circuitbreaker"
+	"github.com/m3db/m3/src/dbnode/generated/thrift/rpc"
+	"github.com/uber-go/tally"
+	tchannel "github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/thrift"
+	"go.uber.org/zap"
+	"sync/atomic"
+)
+
+// circuitBreakerClient is a client that wraps a TChannel client with a circuit breaker.
+type circuitBreakerClient struct {
+	enabled    bool
+	shadowMode bool
+	logger     *zap.Logger
+	cb         *atomic.Value // *circuitbreaker.Circuit
+	metrics    *circuitBreakerMetrics
+	host       string
+	next       rpc.TChanNode
+}
+
+// m3dbMiddleware is a function that takes a TChannel client and returns a circuit breaker client interface.
+type m3dbMiddleware func(rpc.TChanNode) CircuitBreakerClient
+
+// CircuitBreakerClient defines the interface for a circuit breaker client.
+type CircuitBreakerClient interface {
```

Comment:
same point as above about naming: `type Client interface` is probably descriptive enough